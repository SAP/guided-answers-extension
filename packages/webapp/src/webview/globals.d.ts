interface Window {
    vscode: {
        postMessage: (message: string | JSON) => void;
    };
}
