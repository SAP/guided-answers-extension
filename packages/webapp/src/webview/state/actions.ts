export {
    updateGuidedAnswerTrees,
    updateActiveNode,
    selectNode,
    goToPreviousPage,
    goToAllAnswers,
    restartAnswer,
    executeCommand,
    setActiveTree,
    searchTree,
    setQueryValue,
    webviewReady,
    getBetaFeatures,
    sendFeedbackOutcome,
    sendFeedbackComment,
    guideFeedback,
    setProductFilters,
    setComponentFilters,
    resetFilters,
    setPageSize,
    feedbackStatus,
    feedbackResponse
} from '@sap/guided-answers-extension-types';
