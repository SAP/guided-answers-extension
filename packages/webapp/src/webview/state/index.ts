export { actions, store } from './store';
export { getInitialState } from './reducers';
