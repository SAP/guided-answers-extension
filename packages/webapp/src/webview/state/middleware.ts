import i18next from 'i18next';
import type { Middleware, MiddlewareAPI, Dispatch, Action } from 'redux';
import type { GuidedAnswerActions } from '@sap/guided-answers-extension-types';
import type { AppState } from '../types';

declare let window: Window;
declare let acquireVsCodeApi: () => typeof window['vscode'];

/**
 * Communication between IDE extension and web view is realized through the communication middleware
 * window event listener for incoming messages in to the webview
 * post message for outgoing messages to the host.
 *
 * @param store - redux store
 * @returns - redux middleware
 */
export const communicationMiddleware: Middleware<
    Dispatch<GuidedAnswerActions>,
    AppState,
    Dispatch<GuidedAnswerActions>
> = (store: MiddlewareAPI) => {
    // Add event handler, this will dispatch incomming state updates
    window.addEventListener('message', (event: MessageEvent) => {
        if (event.origin === window.origin) {
            console.log(i18next.t('MESSAGE_RECEIVED'), event);
            if (event.data && typeof event.data.type === 'string') {
                store.dispatch(event.data);
            }
        } else {
            console.error(i18next.t('UNKNOWN_ORIGIN') + event.origin);
        }
    });
    try {
        window.vscode = acquireVsCodeApi();
    } catch (e) {
        console.error(i18next.t('ERROR_NOT_RUNNING_IN_WEB_VIEW'));
    }
    return (next: Dispatch<GuidedAnswerActions>) =>
        (action): Action => {
            action = next(action);
            if (action && typeof action.type === 'string' && !action.type.startsWith('[view]')) {
                window.vscode.postMessage(action);
                console.log(i18next.t('REACT_ACTION_POSTED'), action);
            }
            return action;
        };
};
