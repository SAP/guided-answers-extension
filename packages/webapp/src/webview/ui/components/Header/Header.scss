.guided-answer {
    &__header {
        display: flex;
        width: 100%;
        z-index: 10;
        background-color: var(--vscode-editor-background);
        position: sticky;
        top: 0;
        padding-bottom: 20px;
        &__sub {
            display: flex;
            align-items: center;
            width: 100%;
        }
        &__title {
            font-weight: 600;
            margin: 0;
            color: var(--vscode-settings-headerForeground);
            line-height: 29px;
        }
        &__subtitle {
            color: var(--vscode-editorCodeLens-foreground, #999);
            font-size: 2em;
        }
        &__navButtons {
            height: 25px;
            margin-top: 30px;
            color: var(--vscode-foreground);
            padding: 0 5px 0px 5px;
            text-align: start;
            cursor: pointer;
            background: none;
            border: 0;
            &__content {
                vertical-align: middle;
                display: inline-block;
                &__text {
                    margin-left: 3px;
                }
            }
            display: flex;
            align-items: center;
            svg {
                width: 16px;
                height: 16px;
            }
        }
        &__navButtons:hover {
            background: var(--vscode-menubar-selectionBackground);
            outline: 1px solid var(--vscode-contrastActiveBorder);
            border-radius: 3px;
            > span {
                text-decoration: none;
            }
        }
        &__navButtons:focus-visible {
            outline: 1px solid var(--vscode-focusBorder, #007fd4);
        }

        &__allAnswersButton {
            width: 25%;
            margin-right: 80px;
        }
        &__back-restart-buttons {
            display: flex;
            justify-content: space-between;
            width: 75%;
            max-width: 620px;
        }
        &__restart-feedback-buttons {
            display: flex;
        }
        &__divider {
            border-left: 1px solid var(--vscode-editorWidget-border);
            height: 16px;
            margin-top: 35px;
            margin-left: 10px;
            margin-right: 10px;
        }
        &__logoAndTitle {
            display: flex;
            align-items: center;
            margin-top: 30px;
            min-width: 280px;
        }
        &__searchField {
            display: flex;
            align-items: start;
            flex-direction: row;
            margin-top: 30px;
            width: 66%;
        }
    }
}

.home-icon {
    width: 62px;
}

@media only screen and (max-width: 625px) {
    .guided-answer {
        &__header {
            &__logoAndTitle {
                width: 50%;
            }

            &__sub {
                align-items: start;
                flex-direction: column;
            }

            &__searchField {
                margin-top: 20px;
                align-items: start;
            }
            &__navButtons {
                text-align: center;
                justify-content: center;
            }
            &__navButtons__content {
                margin-left: 0;
            }
            &__navButtons__content__text {
                display: none;
            }
            &__back-restart-buttons {
                justify-content: space-between;
                width: 100%;
            }
            &__allAnswersButton {
                margin-right: 48px;
                width: 28px;
                margin-left: 6px;
            }
        }
    }
    .home-icon {
        width: 28px;
        padding: 0 3px 2px 3px;
        border-radius: 3px;
        color: var(--vscode-menu-foreground, #cccccc);
    }
    .home-icon:hover {
        background: var(--vscode-menubar-selectionBackground);
    }
}
