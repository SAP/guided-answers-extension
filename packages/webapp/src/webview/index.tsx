import React from 'react';
import { render } from 'react-dom';
import { Provider } from 'react-redux';
import { App } from './ui/components/App';
import { actions, store } from './state';
import { initI18n } from './i18n';

/**
 * Initialization of i18n, icons
 */
initI18n();

/**
 * Render app
 */
render(
    <Provider store={store}>
        <App />
    </Provider>,
    document.getElementById('root') as HTMLElement
);

/**
 * Fire action which is captured in hosting app (ide-extension) to signal that we are ready to receive messages
 */
actions.webviewReady();

/**Once webview is ready we focus to search field */
document.getElementById('search-field')?.focus();
