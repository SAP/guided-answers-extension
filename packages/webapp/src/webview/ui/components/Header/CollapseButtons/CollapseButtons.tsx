import React from 'react';
import i18next from 'i18next';
import { useSelector } from 'react-redux';
import { UIIconButton, UiIcons } from '@sap-ux/ui-components';
import { actions } from '../../../../state';
import type { AppState } from '../../../../types';

/**
 *
 * @returns Collapse buttons for the header
 */
export function CollapseButtons() {
    const appState = useSelector<AppState, AppState>((state) => state);

    return (
        <>
            <UIIconButton
                id="expand-all-button"
                className="guided-answer__header__navButtons"
                style={{ marginTop: 0 }}
                iconProps={{ iconName: UiIcons.ExpandNodes }}
                title={i18next.t('EXPAND_ALL_TREES')}
                disabled={appState.guidedAnswerTreeSearchResult.resultSize <= 0}
                onClick={() => actions.expandAllSearchNodes()}
                primary></UIIconButton>
            <UIIconButton
                id="collapse-all-button"
                className="guided-answer__header__navButtons"
                style={{ marginTop: 0 }}
                iconProps={{ iconName: UiIcons.ContractNodes }}
                title={i18next.t('COLLAPSE_ALL_TREES')}
                disabled={appState.guidedAnswerTreeSearchResult.resultSize <= 0}
                onClick={() => actions.collapseAllSearchNodes()}
                primary></UIIconButton>
        </>
    );
}
