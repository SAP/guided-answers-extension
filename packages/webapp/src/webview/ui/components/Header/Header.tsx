import React, { ReactElement } from 'react';
import { useSelector } from 'react-redux';
import { AppState } from '../../../types';
import { AllAnswersButton, BackButton, GeneralFeedbackButton, RestartButton } from './NavigationButtons';
import { Logo } from './Logo';
import './Header.scss';
import { SearchField } from './SearchField';
import { FocusZone } from '@fluentui/react-focus';

/**
 * Renders and returns the header section.
 *
 * @param props Props for Header component
 * @param props.showSub Show/Hide subsection of the Header
 * @param props.showNavButons Show/Hide navigation buttons
 * @param props.showLogo Show/Hide SAP logo
 * @param props.showSearch Show/Hide Search field
 * @returns - react element for header section
 */
export function Header(props: {
    showSub: boolean;
    showNavButons: boolean;
    showLogo: boolean;
    showSearch: boolean;
}): ReactElement {
    const appState = useSelector<AppState, AppState>((state) => state);
    return (
        <div className="guided-answer__header" style={{ paddingBottom: props.showNavButons === true ? '0' : '20px' }}>
            {props.showSub === true ? (
                <>
                    <div className="guided-answer__header__sub">
                        {!!props.showLogo && <Logo />}
                        {!!props.showSearch && <SearchField />}
                    </div>{' '}
                </>
            ) : (
                <></>
            )}

            {props.showNavButons === true ? (
                <>
                    <FocusZone isCircularNavigation={true} className="guided-answer__header">
                        <div className="guided-answer__header__allAnswersButton">
                            <AllAnswersButton />
                        </div>

                        <div className="guided-answer__header__back-restart-buttons">
                            <BackButton />
                            <div className="guided-answer__header__restart-feedback-buttons">
                                {appState.activeGuidedAnswerNode.length > 1 && <RestartButton />}
                                {appState.activeGuidedAnswerNode.length > 1 && (
                                    <div className="guided-answer__header__divider"></div>
                                )}
                                <GeneralFeedbackButton />
                            </div>
                        </div>
                    </FocusZone>
                </>
            ) : (
                <></>
            )}
        </div>
    );
}
