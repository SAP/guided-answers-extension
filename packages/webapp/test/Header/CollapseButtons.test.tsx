import React from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import { fireEvent } from '@testing-library/dom';
import { useSelector } from 'react-redux';
import { actions } from '../../src/webview/state';
import { CollapseButtons } from '../../src/webview/ui/components/Header/CollapseButtons';

jest.mock('../../src/webview/state', () => ({
    actions: {
        expandAllSearchNodes: jest.fn(),
        collapseAllSearchNodes: jest.fn()
    }
}));

jest.mock('react-redux', () => ({
    useSelector: jest.fn()
}));

describe('<CollapseButtons />', () => {
    afterEach(cleanup);

    it('renders without crashing', () => {
        (useSelector as jest.Mock).mockImplementation((selector) =>
            selector({
                guidedAnswerTreeSearchResult: {
                    resultSize: 1
                }
            })
        );

        const { container } = render(<CollapseButtons />);
        expect(container).toMatchSnapshot();
    });

    it('renders without crashing, disabled buttons', () => {
        (useSelector as jest.Mock).mockImplementation((selector) =>
            selector({
                guidedAnswerTreeSearchResult: {
                    resultSize: 0
                }
            })
        );

        const { container } = render(<CollapseButtons />);
        expect(container).toMatchSnapshot();
    });

    it('expands all trees', () => {
        (useSelector as jest.Mock).mockImplementation((selector) =>
            selector({
                guidedAnswerTreeSearchResult: {
                    resultSize: 1
                }
            })
        );

        render(<CollapseButtons />);
        fireEvent.click(screen.getByTestId('expand-all-button'));
        expect(actions.expandAllSearchNodes).toBeCalledTimes(1);
    });

    it('collapse all trees', () => {
        (useSelector as jest.Mock).mockImplementation((selector) =>
            selector({
                guidedAnswerTreeSearchResult: {
                    resultSize: 1
                }
            })
        );

        render(<CollapseButtons />);
        fireEvent.click(screen.getByTestId('collapse-all-button'));
        expect(actions.collapseAllSearchNodes).toBeCalledTimes(1);
    });
});
