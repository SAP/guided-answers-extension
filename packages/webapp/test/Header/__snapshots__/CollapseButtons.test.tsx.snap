// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CollapseButtons /> renders without crashing 1`] = `
<div>
  <button
    class="ms-Button ms-Button--icon guided-answer__header__navButtons root-166"
    data-is-focusable="true"
    id="expand-all-button"
    style="margin-top: 0px;"
    type="button"
  >
    <span
      class="ms-Button-flexContainer flexContainer-167"
      data-automationid="splitbuttonprimary"
    >
      <i
        aria-hidden="true"
        class="ms-Icon root-162 ms-Button-icon icon-169"
        data-icon-name="ExpandNodes"
      />
    </span>
  </button>
  <button
    class="ms-Button ms-Button--icon guided-answer__header__navButtons root-166"
    data-is-focusable="true"
    id="collapse-all-button"
    style="margin-top: 0px;"
    type="button"
  >
    <span
      class="ms-Button-flexContainer flexContainer-167"
      data-automationid="splitbuttonprimary"
    >
      <i
        aria-hidden="true"
        class="ms-Icon root-162 ms-Button-icon icon-169"
        data-icon-name="ContractNodes"
      />
    </span>
  </button>
</div>
`;

exports[`<CollapseButtons /> renders without crashing, disabled buttons 1`] = `
<div>
  <button
    aria-disabled="true"
    class="ms-Button ms-Button--icon is-disabled guided-answer__header__navButtons root-174"
    data-is-focusable="false"
    disabled=""
    id="expand-all-button"
    style="margin-top: 0px;"
    type="button"
  >
    <span
      class="ms-Button-flexContainer flexContainer-167"
      data-automationid="splitbuttonprimary"
    >
      <i
        aria-hidden="true"
        class="ms-Icon root-162 ms-Button-icon icon-175"
        data-icon-name="ExpandNodes"
      />
    </span>
  </button>
  <button
    aria-disabled="true"
    class="ms-Button ms-Button--icon is-disabled guided-answer__header__navButtons root-174"
    data-is-focusable="false"
    disabled=""
    id="collapse-all-button"
    style="margin-top: 0px;"
    type="button"
  >
    <span
      class="ms-Button-flexContainer flexContainer-167"
      data-automationid="splitbuttonprimary"
    >
      <i
        aria-hidden="true"
        class="ms-Icon root-162 ms-Button-icon icon-175"
        data-icon-name="ContractNodes"
      />
    </span>
  </button>
</div>
`;
